### ISSUE TITLE
Asynchronous config source

### ISSUE THREAD

**Issue by szarykott** [mehcode/config-rs#148]
I was wondering if there is a broader need for what can be called asnychronous config source?

At my company we use dedicated configuration service that can be asked for config for given application via HTTP. I believe there are also other solutions that use network to retrieve configuration (for instance a database source). 
While implementing `Source` trait for our configuration service I had to use blocking HTTP client. One may say this is not a problem since configuration refresh is not a frequent action. However, when you try to squeeze every ounce of performance from your application or you are using single-thread scheduler this can hurt. Also, in our specific case config is expected to be read by applications in real-time almost, so refresh should be quite frequent.

While implementing blocking HTTP source I encountered a problem with blocking `Client` from `reqwest` crate, since blocking client tries to start a Tokio runtime underneath and use non-blocking implementation, but block on it with `tokio::block_on`. Which basically panics, since if you are using tokio runtime, you are not allowed to set up another runtime inside. There are workarounds for it like:
* push config download out of runtime (but you can have hard time refreshing it then)
* use another library to perform blocking call (which can hurt on single-thread sheduler and makes this crate less friendly to use)

For those reasons I though about asynchronous configuration source.

I am afraid however that it would be hard to make it both easy to use, performant and backwards-compatible. I would drop the latter if I were to choose.

Since I come from .NET world I am quite used to `IConfiguration` introduced by .NET Core. This crate resembles it to some extent.
The idea in refactoring to use asynchoronous source could be (bit similar to `IConfiguration` having its dedicated builder):
* extend `Source` trait with `collect_async` method
* give it a default implementation using `futures::future::ready` with result of `collect` method for ease of use in case of synchronous sources (if possible)
* add `ConfigBuilder` that would **not** refresh config after every merge (might be quite ineffcient to refresh config from HTTP many times in short period of time), instead if would produce config with:
    * `build` method in case you only add synchronous sources (that returns Err if there is even single asynchronous source)
    * `build_async` method in case you have even single asynchronous source (which can build synchronous sources as well)
* library should be kept executor agnostic

What is your take on that? And, **is this crate maintained**? It looks really nice, but last comitts are quite old. I would be happy to make a pull request with aforementioned changes, but I wouldn't like to do work that would hang in a PR for eternity.

**Comment by szarykott** [comment]
@matthiasbeyer since you took ownership of the crate you might be interested in one thing.

Since I did not believe this crate in maintained anymore, I created my own crate with asynchronous configuration source, etc. It follows a bit different paradigm regarding building, but I believe that if people like it some functionalities could be ported to config-rs. There are extensive unit tests. It needs some adjustments but maybe is feasible.

I could try to port some of them, especially regarding asynchronous sources here in my free time. I would not like my work to go to waste as I spent a couple of months creating my crate thinking this one is completely dead without anyone to even merge my PRs.

Tell me what you think. Crate's GitHub is https://github.com/szarykott/miau

**Comment by matthiasbeyer** [comment]
Hi!


>I could try to port some of them, especially regarding asynchronous
>sources here in my free time. I would not like my work to go to waste
>as I spent a couple of months creating my crate thinking this one is
>completely dead without anyone to even merge my PRs.
>
>Tell me what you think. Crate's GitHub is
>https://github.com/szarykott/miau

AWESOME! I'd love to collaborate on this! I'm going on vacation right now, but I'm not off the grid. I might respond slow for the next 7 days, though.

I'll check out your crate! Feel free to draft a PR on integration, if possible! I definitely am interested!

**Comment by szarykott** [comment]
I am glad that you like it!

It will not be fast migration, especially that both libraries follow slightly different patterns, most notably:
* builder: config-rs builds configuration every time there is a new source, which is an inefficient way of doing things if, for instance, a couple of your sources do HTTP requests. Miau builds all sources only once.
* format/source distinction: in config-rs, some sources like file source are a bit tangled with a format.

Nonetheless, I'll try to look into it and prepare some design propositions so that people have time to comment before trying to implement it. Let's not rush!

**Comment by matthiasbeyer** [comment]
No, don't worry about being fast! If we can step-by-step (PR by PR) merge the functionality, that'd be awesome.
I also have some ideas for this crate, most notably in-place reloading and async sources that would be required for etcd support.

All of these things should be, IMHO, compiletime-flags so that users of this crate can decide themselves whether they just want read config files or have more functionality.

On that first point: https://github.com/mehcode/config-rs/pull/180 might be a first step into the direction of "only compute cache once".
I'm not sure what you mean by that second point... but maybe that's a topic for a new issue?

> Let's not rush!

This!

**Comment by szarykott** [comment]
Since #196 has been merged, I'll proceed with this one if it is still desired.

**Comment by matthiasbeyer** [comment]
yes, of course!
